pkg/protowalk/field_path.go

```diff
@@ -0,0 +1,141 @@
+package protowalk
+
+import (
+	"reflect"
+	"strings"
+
+	"github.com/stackrox/rox/pkg/stringutils"
+)
+
+// FieldKind represents the kind of a field.
+type FieldKind int
+
+const (
+	// MessageFieldKind indicates that the element of the respective field is of message type (struct pointer), and
+	// thus can have zero or more child fields.
+	MessageFieldKind FieldKind = iota
+	// OneofFieldKind indicates that the element of the repsective field is a oneof wrapper (interface that is nil
+	// or a struct wrapper), and has exactly one child field.
+	OneofFieldKind
+	// LeafFieldKind indicates that the element of the respective field is neither a message nor oneof, and does not
+	// have any child fields.
+	LeafFieldKind
+)
+
+// Field represents a field in a proto message type or oneof wrapper.
+type Field struct {
+	// ContainingType is a pointer to the message or oneof wrapper type to which this field belongs.
+	// It is always a struct pointer.
+	ContainingType reflect.Type
+	// StructField is the definition of the struct field within the containing type.
+	reflect.StructField
+}
+
+// ElemType is the type of element stored in this field. In case of slices (repeated fields), this is the slice
+// element type; in case of maps, this is the value type. Otherwise, it is the type of the respective struct field.
+func (f Field) ElemType() reflect.Type {
+	ty := f.Type
+	if ty.Kind() == reflect.Slice || ty.Kind() == reflect.Map {
+		ty = ty.Elem()
+	}
+	return ty
+}
+
+// Kind returns the kind of this field.
+func (f Field) Kind() FieldKind {
+	elemTy := f.ElemType()
+	if elemTy.Kind() == reflect.Interface {
+		return OneofFieldKind
+	}
+	if elemTy.Kind() == reflect.Ptr && elemTy.Elem().Kind() == reflect.Struct {
+		return MessageFieldKind
```

Comment:
hmm so one of the perf benefits of gogo proto (which we don't use) was the ability to use fewer pointers. Does vtproto have anything like that?